components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/GatewayArtifactsMgtDBUtil.java

```diff
@@ -0,0 +1,69 @@
+package org.wso2.carbon.apimgt.impl.utils;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.apimgt.api.APIManagementException;
+import org.wso2.carbon.apimgt.api.APIManagerDatabaseException;
+import org.wso2.carbon.apimgt.impl.dto.GatewayArtifactSynchronizerProperties;
+import org.wso2.carbon.apimgt.impl.internal.ServiceReferenceHolder;
+import javax.naming.Context;
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.sql.SQLException;
+
+public class GatewayArtifactsMgtDBUtil {
+
+    private static final Log log = LogFactory.getLog(GatewayArtifactsMgtDBUtil.class);
+    private static volatile DataSource artifactSynchronizerDataSource = null;
+
+    /**
+     * Initializes the data source
+     *
+     * @throws APIManagementException if an error occurs while loading DB configuration
+     */
+    public static void initialize() throws APIManagerDatabaseException {
+        if (artifactSynchronizerDataSource != null) {
+            return;
+        }
+
+        synchronized (GatewayArtifactsMgtDBUtil.class) {
```

Comment:
Instead of synchronizing the entire class(which leads to locking all methods of the class), it would be better to move this block into a private synchronized method. It would then look like,

```
public static void initialize() throws APIManagerDatabaseException {
        if (artifactSynchronizerDataSource != null) {
            return;
        }

        InitDataSource();
}

private static void InitDataSource() synchronized {
    .............
}
```

This would ensure that only callers of the `initialize()` method will have the potential of having to wait in a concurrent scenario, but callers of the `getArtifactSynchronizerConnection()` method will not be affected by this.

Reply:
https://github.com/wso2/carbon-apimgt/pull/8828/commits/dab832307b308f4cee423710218258ee6b0ed015